lib/ae_mdw_web/websocket/broadcaster.ex

```diff
@@ -109,111 +116,72 @@ defmodule AeMdwWeb.Websocket.Broadcaster do
-    msg =
-      header
-      |> :aec_headers.serialize_for_client(prev_block_type)
-      |> encode_message("KeyBlocks", source)
-
-    broadcast("KeyBlocks", version, msg)
+  defp serialize_block(header, :key, :mdw, :v2) do
+    height = header |> :aec_headers.height() |> Integer.to_string()
+    Blocks.fetch_key_block(State.mem_state(), height)
   end
 
-  defp do_broadcast_micro_block(header, :v2, :mdw) do
+  defp serialize_block(header, :micro, :mdw, :v2) do
     {:ok, hash} = :aec_headers.hash_header(header)
-    state = State.mem_state()
-
-    case Blocks.fetch_micro_block(state, hash) do
-      {:ok, block} ->
-        msg = encode_message(block, "MicroBlocks", block)
-        broadcast("MicroBlocks", :v2, msg)
-
-      {:error, _reason} ->
-        :ok
-    end
+    Blocks.fetch_micro_block(State.mem_state(), encode(:micro_block_hash, hash))
   end
 
-  defp do_broadcast_micro_block(header, version, source) do
+  defp serialize_block(header, _type, _source, _version) do
     prev_block_type = Db.prev_block_type(header)
-
-    msg =
-      header
-      |> :aec_headers.serialize_for_client(prev_block_type)
-      |> encode_message("MicroBlocks", source)
-
-    broadcast("MicroBlocks", version, msg)
+    {:ok, :aec_headers.serialize_for_client(header, prev_block_type)}
   end
 
-  defp do_broadcast_txs(block, :v2, :mdw) do
-    state = State.mem_state()
-    tx_pids = Subscriptions.subscribers(:v2, "Transactions")
+  defp do_broadcast_txs(block, source, version) do
+    tx_pids = Subscriptions.subscribers(source, version, "Transactions")
+    context = if {source, version} == {:mdw, :v2}, do: State.mem_state(), else: block
 
     block
     |> :aec_blocks.txs()
     |> Enum.each(fn tx ->
-      tx_hash = :aetx_sign.hash(tx)
+      with {:ok, mdw_tx} <- serialize_tx(tx, context) do
+        mdw_tx
+        |> encode_message("Transactions", source)
+        |> broadcast_tx(tx_pids)
 
-      case Txs.fetch(state, tx_hash, true) do
-        {:ok, mdw_tx} ->
-          mdw_tx
-          |> encode_message("Transactions", :mdw)
-          |> broadcast_tx(tx_pids)
+        obj_msg = encode_message(mdw_tx, "Object", source)
 
-          obj_msg = encode_message(mdw_tx, "Object", :mdw)
-
-          tx
-          |> get_ids_from_tx()
-          |> Enum.each(&broadcast(&1, :v2, obj_msg))
-
-        {:error, _reason} ->
-          :ok
+        tx
+        |> get_ids_from_tx()
+        |> Enum.each(&broadcast(obj_msg, &1, source, version))
       end
     end)
   end
 
-  defp do_broadcast_txs(block, version, source) do
-    header = :aec_blocks.to_header(block)
-    tx_pids = Subscriptions.subscribers(version, "Transactions")
-
-    block
-    |> :aec_blocks.txs()
-    |> Enum.each(fn tx ->
-      ser_tx = :aetx_sign.serialize_for_client(header, tx)
-
-      ser_tx
-      |> encode_message("Transactions", source)
-      |> broadcast_tx(tx_pids)
-
-      obj_msg = encode_message(ser_tx, "Object", source)
+  defp serialize_tx(tx, %State{} = state) do
```

Comment:
We can instead tag terms to avoid looking into opaque term structures. E.g. `{:state, state} | {:block, block}`

Reply:
It's not looking to data, only to struct type but refactored to help your best friend dialyzer.

Reply:
Thanks